RocketChat/Rocket.Chat
# feat: Allow agents to forward to offiline department/agents

## Proposed changes (including videos or screenshots)

This pull request introduces a pivotal enhancement, empowering live chat agents to efficiently manage inquiries directed to offline departments.

Central to this upgrade is the incorporation of the Livechat_allow_forward_inquiry_to_offline_department_agents setting. This setting seamlessly facilitates the forwarding of inquiries, ensuring offline agents are seamlessly integrated into the process.

Furthermore, to optimize maintainability and readability, extensive method refactoring has been conducted. Specifically, the logic for room.forward has been relocated to a separate folder. Additionally, the forwardRoomToAgent and forwardRoomToDepartment methods have been revamped to direct all forwarding information to the Queue Manager. This strategic adjustment eliminates unnecessary logic, streamlining the forwarding process. Consequently, the forwarding logic is now exclusively managed by the Queue Manager, while the forwards method exclusively determines forwarding recipients based on the offline setting's logic.

## Issue(s)

[SUP-242](https://rocketchat.atlassian.net/browse/SUP-242)

## Steps to test or reproduce

1. Set the mentioned setting as true
2. Initiate a conversation on the live chat widget
3. Forward the conversation to an offline agent/department
4. Ensure the visitor is transferred to the selected department/agent

## Further comments

Developer notes:

- The Room information is not reactive; when the user is forwarded, the information remains the same in the contextual bar.
- Some other parts of the live chat need a refactor as well; there are many methods with too many lines, containing verifications that may not be needed anymore.
- Attempted to add unit tests for the refactor methods, but the complexity of mocking the methods for testing could lead to a frustrating journey within a limited timeframe.

[SUP-242]: https://rocketchat.atlassian.net/browse/SUP-242?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
